trunk/public/class-facebook-login-public.php

```diff
@@ -212,8 +237,22 @@ public function jetpack_script_args( $args ){
 	public function login_or_register_user() {
 		check_ajax_referer( 'facebook-nonce', 'security' );
 
-		$access_token = isset( $_POST['fb_response']['authResponse']['accessToken'] ) ? $_POST['fb_response']['authResponse']['accessToken'] : '';
-		$fb_user_id = $_POST['fb_response']['authResponse']['userID'];
+		$redirect = home_url();
+		$parsed_home_url = parse_url( home_url() );
+        if( ! empty( $_GET['redirect_to'] ) ) {
+			$parsed_get_url = parse_url( $_GET['redirect_to'] );
+			if( $parsed_get_url['host'] === $parsed_home_url['host'] ) {
+				$redirect = esc_url( $_GET['redirect_to'] );
+			}
+		} else if( ! empty( $_SERVER['HTTP_REFERER'] ) ) {
```

Comment:
HTTP_REFERER is not safe, I know we're checking it but is there any reason to check for it? IN case a user is logging from a page other than the homepage, passing redirect_to should redirect appropriately.

Reply:
In this block, the elseif is getting executed, not the if.

```$redirect = home_url();
		$parsed_home_url = parse_url( home_url() );
        if( ! empty( $_GET['redirect_to'] ) ) {
			$parsed_get_url = parse_url( $_GET['redirect_to'] );
			if( $parsed_get_url['host'] === $parsed_home_url['host'] ) {
				$redirect = esc_url( $_GET['redirect_to'] );
			}
		} else if( ! empty( $_SERVER['HTTP_REFERER'] ) ) {
			$parsed_referer = parse_url( $_SERVER['HTTP_REFERER'] );
			if( $parsed_referer['host'] === $parsed_home_url['host'] ) {
				$redirect = esc_url( $_SERVER['HTTP_REFERER'] );
			}
		}```

 So it seems like http_referer is needed as it's written - redirect_to is passed if a user tries to access a page that's only available to logged in users, such as wp-admin. It would be better to only use $redirect_to, though. Maybe that's why $_GET['redirect'] was being used before, to pass it to Facebook. I'll see if I can get it to pass redirect_to instead.

Reply:
I passed the referer to js as $wp->request. That ought to be safe.